sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/FieldValueGetter.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.values.reflect.field;
+
+import org.apache.beam.sdk.values.reflect.RecordFactory;
+
+/**
+ * An interface to access a field of a class.
+ *
+ * <p>Implementations of this interface are generated at rutime by {@link RecordFactory}
+ * to map pojo fields to BeamRecord fields.
+ */
+public interface FieldValueGetter<T> {
```

Comment:
I am a bit puzzled by this Interface, is this really needed ? I mean with the Pojo object we have all that is needed,  we have the name/type by reflection and the value just by invoking the method.

Reply:
If you use reflection, then you probably don't need this specific interface. But I don't think that getting rid of it, or switching to reflection will reduce complexity a lot without sacrificing something.

This is how I understand it:

Reflection or other implementation details aside, there are 2 problems:
 - map a pojo class to a record type;
 - given a pojo instance, create a record from it, populating the fields in the order they are specified in the record type;

Creating a record type is straightforward, you need to map pojo fields/getters to record field names and coders. You can do it either by convention or using an explicit spec. Record types can be either cached for pojo classes, or re-created for every pojo instance. Overall this feels straightforward. 

Second problem creating a record from a pojo conceptually looks like this:
 - look up or create the record type for the pojo class;
 - get the fields in correct order (same as in record type);
 - for each field invoke the corresponding getter on the pojo;

This means that now we need a mechanism to map record field types to pojo getters (here I assume that even with reflection we will still need to cache `Method`s).

Given the above you seem to need few abstractions no matter whether you use reflection or not:
- [main component](https://github.com/apache/beam/pull/4204/files#diff-55e6442c81f404c1004a445b550f03c9) to create records based on pojos which will:
   - get the record type for a pojo class;
   - while we are creating a record type we know which fields order we're using and which getters correspond to those, so we are also storing a [mapping of ordered list of getters to corresponding record type](https://github.com/apache/beam/pull/4204/files#diff-9b4f7bc45032683f6e0e7d12294057d8);
   - invoke pojo getters in correct order corresponding to record type;
-  component to create record types, which will:
   - map getter names to record field names;
   - map getter return types to coders;
   - there will be 2 flavors of this ([one](https://github.com/apache/beam/pull/4204/files#diff-8910899b54115ac722de323709421db6), [two](https://github.com/apache/beam/pull/4204/files#diff-271ec0f18add17fe3046ec62984ebc89)) because SQL has its custom coders mapping;

So I feel like most of the components in this PR will be needed in some way or another, and the main question is how you map getters to fields and invoke them.

Using reflection:
 - store each getter in a `List<Method>`;
 - invoke a getter with reflection;

Using code generation:
 - generate a class to represent a `Method` call;
 - invoke getter the class directly;

There are pros and cons for both. How I see it:

Reflection:
 - simpler code because you don't have the code generation logic and custom class to represent the getter method;
 - the implementation is java runtime specific on multiple levels;
 - feels like it might require more serialization code if we want to pass the schema information around;
 - easy to shoot yourself in the foot if you start doing more reflection than needed;

Code generation:
 - more code;
 - potentially can be optimized further by JIT;
 - doesn't have to be at run-time;
 - similar interfaces / mechanisms can be implemented in other languages for schema support;

Reply:
Thanks for specifying this, I understand now better the motivation.sorry I forgot to send some points after checking this, I think we agree in the core ideas/goal, my only complain is more API based. Some notes inline:

> Reflection or other implementation details aside, there are 2 problems:
> 
>     map a pojo class to a record type;
>     given a pojo instance, create a record from it, populating the fields in the order they are specified in the record type;

Sure !

You can quickly achieve this by mapping the POJO class into a BeamRecord type. This is just getting the public get* methods and create a LinkedHashMap<String, Coder> where the key is the name of the field and the coder is the given coder for the type.

Caching this mapping seems like an excellent idea because this is a time consuming step. So it should be cached in some sort of Map<Class, BeamRecordType> where the Key is the Pojo Class.

> Second problem creating a record from a pojo conceptually looks like this:
> 
>     look up or create the record type for the pojo class;
>     get the fields in correct order (same as in record type);
>     for each field invoke the corresponding getter on the pojo;
> 
> This means that now we need a mechanism to map record field types to pojo getters (here I assume that even with reflection we will still need to cache Methods).
> 
> Given the above you seem to need few abstractions no matter whether you use reflection or not:

I disagree, you don’t really need extra abstractions for this if you do it by reflection, but more important that needing or not these abstractions is the fact of exposing them in the API. You can do this by simply having a method with a signature like this:

	List<Object> getValues(Object myPojo, List<String> attributes);

and then with that List of values and the BeamRecordType you can build the BeamRecord.

The way you implement this method to make it faster is another story and it is great if we can add some tricks to improve this, but we should try to avoid the optimization from leaking into the API. We are doing this in sdks/java/core, trying not to leak abstractions at this level is important because if we don’t take care here they will be used everywhere.

Reply:
I am not sure I understand your point about exposing the abstractions in the API. The only thing I am planning to expose is `BeamRecordFactory.create(pojo)`. This currently creates the `BeamRecord`s using all public getters from the pojo instances. This PR scope is to only implement this.
Probably eventually there will be another parameter added to specify a schema. Whether it's backed by reflection or code generation should not be leaked at this point. Actual schema support or even getting the `BeamRecord` for a subset of the getters of a pojo is out of the scope here.

Imagine that it is plugged into SQL to convert from pojos to records somewhat like this:

```
class PojoToRecord extends DoFn<Pojo, BeamRecord>() {
  @ProcessElement
  public void processElement(Context c) { 
     c.output(BeamRecordFactory.instance().create(c.element()));
  }
```

> `List<Object> getValues(Object myPojo, List<String> attributes);`

Yes, I agree. Something like this is needed. But to actually create a record you also need to make sure that the `attributes` here and the `BeamRecordType` have the same attributes in the same order. 

To achieve that, it would make sense to store that order somewhere, like `Map<Class, List<String>>`, and construct it at the same time you create the `BeamRecordType` because you're analyzing getters anyway. This way the list of attributes is available in the map when you are creating the beam records later. Then you use this list of attributes to get the values by calling `getValues(myPojo, attributes)`. 

I agree it's all pretty straightforward here, but I argue that it's roughly what's happening here anyway, minus the actual code generation bit.

For example, consider following:
- you can combine the values from `Map<Class, BeamRecordType>` and `Map<Class, List<String>>`  and put them in the utility class: `RecordTypeGetters { beamRecordType, attributes }`, and have a single map `Map<Class, RecordTypeGetters>`.
- when calling `getValues(myPojo, attributes)` the most expensive part is actually looking up the methods on the class, so I would cache them:
  - instead of storing the attribute names, just store the getter methods in a `List<Method>`;
  - so now you can change `getValues(Object myPojo, List<String> attributes)` to `getValues(Object myPojo, List<Method> getters)`;
  - now something like this is already in `BeamRecordFactory`: `List<Object> getFieldValues(List<FieldValueGetter> fieldValueGetters, Object pojo)`;
- you still need to wrap everything into some utility/factory class:
  - getting the getters from pojo;
  - mapping getter return types to coders for sql and non-sql versions;
  - caching;
  - getting the values from pojo;
 - now if you replace super low-level `Method` with our own class `FieldValueGetter` you have my implementation here. The only thing different is whether we get `FieldValueGetter`/`Method` by reflection or by codegen;